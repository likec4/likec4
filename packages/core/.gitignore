# Build outputs
dist/
lib/
*.tgz
builder/
compute-view/
model/
styles/
types/
utils/

# Auto-generated by likec4ops prepack
builder/package.json
compute-view/relationships/package.json
compute-view/package.json
model/package.json
styles/package.json
types/package.json
utils/package.json
