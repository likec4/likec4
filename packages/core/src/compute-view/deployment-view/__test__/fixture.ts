import { map, prop } from 'remeda'
import type { DeploymentRulesBuilderOp } from '../../../builder'
import { Builder } from '../../../builder'
import { LikeC4Model } from '../../../model'
import type { DeploymentView } from '../../../types'
import { withReadableEdges } from '../../utils/with-readable-edges'
import { computeDeploymentView } from '../compute'

const {
  builder: b,
  model: {
    model,
    person,
    system,
    component,
    webapp,
    mobile,
    ...$m
  },
  deployment: {
    deployment,
    env,
    node,
    zone,
    instanceOf,
    ...$d
  },
  views: {
    views,
    deploymentView,
    $rules,
    $include,
    $exclude,
  },
} = Builder.forSpecification({
  elements: {
    person: {
      style: {
        shape: 'person',
      },
    },
    system: {},
    component: {},
    webapp: {
      style: {
        shape: 'browser',
      },
    },
    mobile: {
      style: {
        shape: 'mobile',
      },
    },
  },
  deployments: {
    env: {},
    zone: {},
    node: {},
  },
  tags: ['old', 'next', 'temp'],
  relationships: { 'https': {} },
})

export const builder = b
  .with(
    model(
      person('customer', {
        title: 'Happy Customer',
      }),
      system('cloud').with(
        component('frontend').with(
          webapp('dashboard'),
          mobile('mobile'),
        ),
        component('auth'),
        component('backend').with(
          component('api'),
        ),
        component('media', {
          shape: 'storage',
        }),
      ),
      system('aws').with(
        component('rds', {
          shape: 'storage',
        }),
        component('s3', {
          shape: 'storage',
        }),
      ),
      system('email'),
    ),
    model(
      $m.rel('customer', 'cloud', 'uses services'),
      $m.rel('customer', 'cloud.frontend.mobile', 'opens mobile app'),
      $m.rel('customer', 'cloud.frontend.dashboard', {
        title: 'opens in browser',
        tags: ['old'],
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.auth', {
        title: 'authenticates',
        color: 'green',
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.backend.api', {
        title: 'fetches data',
        tags: ['old'],
      }),
      $m.rel('cloud.frontend.dashboard', 'cloud.media', {
        title: 'fetches media',
        tags: ['temp'],
      }),
      $m.rel('cloud.frontend.mobile', 'cloud.auth', {
        title: 'authenticates',
        color: 'amber',
      }),
      $m.rel('cloud.frontend.mobile', 'cloud.backend.api', 'fetches data'),
      $m.rel('cloud.frontend.mobile', 'cloud.media', 'fetches media'),
      $m.rel('cloud.backend.api', 'cloud.auth', 'authorizes'),
      $m.rel('cloud.backend.api', 'cloud.media', 'uploads media'),
      $m.rel('cloud.backend.api', 'aws.rds', 'reads/writes'),
      $m.rel('cloud.backend.api', 'email', 'sends emails'),
      $m.rel('cloud.media', 'aws.s3', 'uploads'),
      $m.rel('email', 'customer', 'sends emails'),
    ),
    deployment(
      node('customer').with(
        instanceOf('instance', 'customer'),
      ),
      env('prod').with(
        zone('eu').with(
          zone('zone1').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          zone('zone2').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          instanceOf('auth', 'cloud.auth'),
          instanceOf('media', 'cloud.media'),
          instanceOf('db', 'aws.rds'),
        ),
        zone('us').with(
          zone('zone1').with(
            instanceOf('api', 'cloud.backend.api'),
            instanceOf('ui', 'cloud.frontend.dashboard'),
          ),
          instanceOf('db', 'aws.rds'),
        ),
      ),
      env('dev').with(
        node('devCustomer').with(
          instanceOf('instance', 'customer'),
        ),
        node('devCloud').with(
          instanceOf('instance', 'cloud'),
        ),
      ),
      env('acc').with(
        node('testCustomer').with(
          instanceOf('instance', 'customer'),
        ),
        zone('eu').with(
          instanceOf('api', 'cloud.backend.api'),
          instanceOf('ui', 'cloud.frontend.dashboard'),
          instanceOf('auth', 'cloud.auth'),
          instanceOf('db', 'aws.rds'),
        ),
      ),
      env('global').with(
        instanceOf('email', 'email'),
      ),
      $d.rel('prod.eu.db', 'prod.us.db', {
        title: 'replicates',
        color: 'green',
      }),
      $d.rel('prod.us.db', 'prod.eu.db', {
        title: 'replicates',
        color: 'amber',
      }),
    ),
  )
  .clone()

export type Types = typeof builder['Types']

export { $exclude, $include }

export function computeView(...rules: DeploymentRulesBuilderOp<Types>[]) {
  const modelsource = builder.clone()
    .with(
      views(
        deploymentView('index', $rules(...rules)),
      ),
    ).build()

  const model = LikeC4Model.create({
    ...modelsource,
    views: {},
  })

  const view = withReadableEdges(computeDeploymentView(model, modelsource.views.index as DeploymentView))

  return Object.assign(view, {
    nodeIds: map(view.nodes, prop('id')) as string[],
    edgeIds: map(view.edges, prop('id')) as string[],
  })
}

export function computeNodesAndEdges(...rules: DeploymentRulesBuilderOp<Types>[]) {
  const { nodeIds, edgeIds } = computeView(...rules)
  return {
    // Starts with capital letter to be first in snapshot
    Nodes: nodeIds,
    edges: edgeIds,
  }
}

export function createModel() {
  const modelsource = builder.build()

  return LikeC4Model.create({
    ...modelsource,
    views: {},
  })
}
