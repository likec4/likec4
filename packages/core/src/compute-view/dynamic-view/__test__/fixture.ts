import { partition } from 'remeda'
import type {
  DynamicViewIncludeRule,
  DynamicViewRule,
  DynamicViewStep,
  ElementViewPredicate,
  Fqn,
  ParsedDynamicView as DynamicView,
  ViewId,
} from '../../../types'
import { type $Aux, type FakeElementIds, fakeModel } from '../../element-view/__test__/fixture'
import { computeDynamicView } from '../compute'

const emptyView = {
  _type: 'dynamic' as const,
  id: 'index' as ViewId,
  title: null,
  description: null,
  tags: null,
  links: null,
  rules: [],
}

type StepExpr = `${FakeElementIds} ${'->' | '<-'} ${FakeElementIds}`
type StepProps = Omit<DynamicViewStep, 'source' | 'target' | 'isBackward'>

export function $step(expr: StepExpr, props?: string | Partial<StepProps>): DynamicViewStep {
  const title = typeof props === 'string' ? props : props?.title
  if (expr.includes(' -> ')) {
    const [source, target] = expr.split(' -> ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      ...(typeof props === 'object' ? props : {}),
      title: title ?? null,
    }
  }
  if (expr.includes(' <- ')) {
    const [target, source] = expr.split(' <- ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      ...(typeof props === 'object' ? props : {}),
      title: title ?? null,
      isBackward: true,
    }
  }
  throw new Error(`Invalid step expression: ${expr}`)
}

export function compute(
  stepsAndRules: (DynamicViewStep<$Aux> | ElementViewPredicate<$Aux> | DynamicViewIncludeRule<$Aux>)[],
) {
  const [steps, rules] = partition(stepsAndRules, (s): s is DynamicViewStep => 'source' in s)
  let view = computeDynamicView(
    fakeModel,
    {
      ...emptyView,
      steps,
      rules: rules as DynamicViewRule[],
    } as DynamicView,
  )
  return Object.assign(view, {
    nodeIds: view.nodes.map((node) => node.id) as string[],
    edgeIds: view.edges.map((edge) => edge.id) as string[],
  })
}
