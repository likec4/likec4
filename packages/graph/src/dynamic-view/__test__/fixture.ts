import type { DynamicViewRule, DynamicViewStep, Fqn, ViewID, ViewRuleExpression } from '@likec4/core'
import { partition } from 'remeda'
import { type FakeElementIds, fakeModel } from '../../compute-view/__test__/fixture'
import { computeDynamicView } from '../index'

const emptyView = {
  __: 'dynamic' as const,
  id: 'index' as ViewID,
  title: null,
  description: null,
  tags: null,
  links: null,
  rules: []
}

type StepExpr = `${FakeElementIds} ${'->' | '<-'} ${FakeElementIds}`

export function $step(expr: StepExpr, title?: string): DynamicViewStep {
  if (expr.includes(' -> ')) {
    const [source, target] = expr.split(' -> ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      title: title ?? null
    }
  }
  if (expr.includes(' <- ')) {
    const [target, source] = expr.split(' <- ')
    return {
      source: source as Fqn,
      target: target as Fqn,
      title: title ?? null,
      isBackward: true
    }
  }
  throw new Error(`Invalid step expression: ${expr}`)
}

export function compute(stepsAndRules: (DynamicViewStep | ViewRuleExpression)[]) {
  const [steps, rules] = partition(stepsAndRules, (s): s is DynamicViewStep => 'source' in s)
  const result = computeDynamicView(
    {
      ...emptyView,
      steps,
      rules: rules as DynamicViewRule[]
    },
    fakeModel
  )
  if (!result.isSuccess) {
    throw result.error
  }
  return Object.assign(result.view, {
    nodeIds: result.view.nodes.map((node) => node.id) as string[],
    edgeIds: result.view.edges.map((edge) => edge.id) as string[]
  })
}
