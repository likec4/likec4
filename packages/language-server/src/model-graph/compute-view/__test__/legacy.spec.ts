import { type ElementKind, type Fqn, type IconUrl, type Tag } from '@likec4/core'
import { describe, expect, it } from 'vitest'
import { $exclude, $include, computeView } from './fixture'

describe('compute-element-view', () => {
  it('should be empty if no root and no rules', () => {
    const { nodes, edges } = computeView([])
    expect(nodes).toEqual([])
    expect(edges).toEqual([])
  })

  it('should show only root if no rules', () => {
    const { nodeIds, edges } = computeView('cloud', [])
    expect(nodeIds).toEqual(['cloud'])
    expect(edges).toEqual([])
  })

  it('should return landscape view on top `include *`', () => {
    const { nodes, nodeIds, edgeIds } = computeView([{ include: [{ wildcard: true }] }])
    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'email',
      'amazon'
    ])
    expect(edgeIds).toEqual([
      'customer:cloud',
      'support:cloud',
      'cloud:amazon',
      'cloud:email',
      'email:cloud'
    ])
    const [customer, support, cloud, email, amazon] = nodes
    expect(amazon).toMatchObject({
      outEdges: [],
      inEdges: ['cloud:amazon']
    })
    expect(cloud).toMatchObject({
      outEdges: [
        'cloud:amazon',
        'cloud:email'
      ],
      inEdges: [
        'email:cloud',
        'support:cloud',
        'customer:cloud'
      ]
    })
    expect(customer).toMatchObject({
      outEdges: ['customer:cloud'],
      inEdges: []
    })
    expect(support).toMatchObject({
      outEdges: ['support:cloud'],
      inEdges: []
    })
  })

  // TODO investigate why this test fails
  it.fails('should return nodes in the same order as was in view', () => {
    const { nodeIds, edgeIds } = computeView([$include('email'), $include('*')])
    expect(nodeIds).toEqual([
      'email',
      'customer',
      'support',
      'cloud',
      'amazon'
    ])
    expect(edgeIds).toEqual(['customer:cloud', 'support:cloud', 'cloud:amazon'])
  })

  it('view of cloud', () => {
    const { nodeIds, edgeIds, ...view } = computeView('cloud', [$include('*')])
    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'cloud.frontend',
      'cloud.backend',
      'email',
      'amazon'
    ])

    expect(edgeIds).to.have.same.members([
      'customer:cloud.frontend',
      'support:cloud.frontend',
      'cloud.frontend:cloud.backend',
      'cloud.backend:amazon',
      'cloud.backend:email'
    ])

    expect(view).toMatchSnapshot()
  })

  it('view of cloud.backend', () => {
    const { nodeIds, edgeIds } = computeView('cloud.backend', [$include('*'), $include('customer')])

    expect(nodeIds).toEqual([
      'customer',
      'cloud.frontend',
      'cloud.backend',
      'cloud.backend.graphql',
      'cloud.backend.storage',
      'amazon'
    ])

    expect(edgeIds).to.have.same.members([
      'customer:cloud.frontend',
      'cloud.frontend:cloud.backend.graphql',
      'cloud.backend.graphql:cloud.backend.storage',
      'cloud.backend.storage:amazon'
    ])
  })

  it('view of cloud.frontend', () => {
    const { nodeIds, edgeIds, ...view } = computeView('cloud.frontend', [$include('*')])
    expect(nodeIds).toEqual([
      'support',
      'customer',
      'cloud.frontend',
      'cloud.frontend.adminPanel',
      'cloud.frontend.dashboard',
      'cloud.backend'
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.frontend.adminPanel:cloud.backend',
      'cloud.frontend.dashboard:cloud.backend',
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.adminPanel'
    ])

    expect(view).toMatchSnapshot()
  })

  it('view of amazon', () => {
    const { nodeIds, edgeIds } = computeView('amazon', [$include('*')])

    expect(nodeIds).toEqual(['cloud', 'amazon', 'amazon.s3'])
    expect(edgeIds).toEqual(['cloud:amazon.s3'])
  })

  it('view of amazon including cloud.* ->', () => {
    const { edgeIds, nodeIds } = computeView('amazon', [$include('*'), $include('cloud.* ->')])

    expect(nodeIds).toEqual(['cloud.backend', 'amazon', 'amazon.s3'])

    expect(edgeIds).to.have.same.members(['cloud.backend:amazon.s3'])
  })

  it('view of cloud.frontend (and include parent cloud)', () => {
    const { edgeIds, nodeIds } = computeView('cloud.frontend', [$include('*'), $include('cloud')])

    expect(nodeIds).toEqual([
      'support',
      'customer',
      'cloud',
      'cloud.frontend',
      'cloud.frontend.adminPanel',
      'cloud.frontend.dashboard',
      'cloud.backend'
    ])

    expect(edgeIds).to.have.same.members([
      'support:cloud.frontend.adminPanel',
      'customer:cloud.frontend.dashboard',
      'cloud.frontend.adminPanel:cloud.backend',
      'cloud.frontend.dashboard:cloud.backend'
    ])
  })

  it('view of cloud.frontend (and include cloud.backend.*)', () => {
    const { edgeIds, nodeIds } = computeView('cloud.frontend', [
      $include('*'),
      $include('cloud'),
      $include('cloud.backend.*')
    ])

    expect(nodeIds).toEqual([
      'support',
      'customer',
      'cloud',
      'cloud.frontend',
      'cloud.frontend.adminPanel',
      'cloud.frontend.dashboard',
      'cloud.backend.graphql',
      'cloud.backend.storage'
    ])

    expect(edgeIds).toEqual([
      'support:cloud.frontend.adminPanel',
      'customer:cloud.frontend.dashboard',
      'cloud.frontend.adminPanel:cloud.backend.graphql',
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.backend.graphql:cloud.backend.storage'
    ])
  })

  it('view of cloud.frontend (and include cloud.frontend -> cloud.backend.*)', () => {
    const { edgeIds, nodeIds } = computeView('cloud.frontend', [
      $include('*'),
      $include('cloud'),
      $include('cloud.frontend -> cloud.backend.*')
    ])

    expect(nodeIds).toEqual([
      'support',
      'customer',
      'cloud',
      'cloud.frontend',
      'cloud.frontend.adminPanel',
      'cloud.frontend.dashboard',
      'cloud.backend.graphql'
    ])

    expect(edgeIds).to.have.same.members([
      'support:cloud.frontend.adminPanel',
      'customer:cloud.frontend.dashboard',
      'cloud.frontend.adminPanel:cloud.backend.graphql',
      'cloud.frontend.dashboard:cloud.backend.graphql'
    ])
  })

  it('view of cloud (exclude cloud, amazon)', () => {
    const { edgeIds, nodeIds } = computeView('cloud', [
      $include('*'),
      $exclude('cloud'),
      $exclude('email'),
      $exclude('amazon')
    ])

    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud.frontend',
      'cloud.backend'
    ])

    expect(edgeIds).toEqual([
      'customer:cloud.frontend',
      'support:cloud.frontend',
      'cloud.frontend:cloud.backend'
    ])
  })

  it('view with 3 levels', () => {
    const { edgeIds, nodeIds } = computeView('cloud', [
      $include('*'),
      $include('cloud.frontend.*'),
      $include('cloud.backend.*'),
      $include('-> amazon.*'),
      $exclude('amazon'),
      $exclude('email'),
      $exclude('cloud.frontend')
    ])

    expect(nodeIds).toEqual([
      'support',
      'customer',
      'cloud',
      'cloud.frontend.adminPanel',
      'cloud.frontend.dashboard',
      'cloud.backend',
      'cloud.backend.graphql',
      'cloud.backend.storage',
      'amazon.s3'
    ])

    expect(edgeIds).toEqual([
      'support:cloud.frontend.adminPanel',
      'customer:cloud.frontend.dashboard',
      'cloud.frontend.adminPanel:cloud.backend.graphql',
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.backend.graphql:cloud.backend.storage',
      'cloud.backend.storage:amazon.s3'
    ])
  })

  it('view with 3 levels (with only relevant elements)', () => {
    const { edgeIds, nodeIds } = computeView('cloud', [
      $include('*'),
      $include('-> cloud.frontend.*'),
      $include('-> cloud.backend.*')
    ])

    expect(nodeIds).toEqual([
      'customer',
      'support',
      'cloud',
      'cloud.frontend',
      'cloud.frontend.dashboard',
      'cloud.frontend.adminPanel',
      'cloud.backend',
      'email',
      'amazon',
      'cloud.backend.graphql'
    ])

    expect(edgeIds).toEqual([
      'customer:cloud.frontend.dashboard',
      'support:cloud.frontend.adminPanel',
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.frontend.adminPanel:cloud.backend.graphql',
      'cloud.backend:amazon',
      'cloud.backend:email'
    ])
  })

  it('index view with applied styles', () => {
    const { nodes } = computeView([
      $include('customer'),
      $include('amazon'),
      $include('cloud'),
      $include('cloud.frontend.dashboard'),
      // all elements
      // color: secondary
      {
        targets: [{ wildcard: true }],
        style: {
          color: 'secondary',
          icon: 'none',
          shape: 'storage'
        }
      },
      // cloud
      // color: muted
      {
        targets: [{ element: 'cloud' as Fqn }],
        style: {
          color: 'muted',
          icon: 'http://some-icon' as IconUrl
        }
      },
      // cloud.*
      // shape: browser
      {
        targets: [{ element: 'cloud' as Fqn, isDescedants: true }],
        style: {
          shape: 'browser'
        }
      }
    ])

    const amazon = nodes.find(n => n.id === 'amazon')!
    const customer = nodes.find(n => n.id === 'customer')!
    const cloud = nodes.find(n => n.id === 'cloud')!
    const frontend = nodes.find(n => n.id === 'cloud.frontend.dashboard')!

    expect(amazon).toMatchObject({
      color: 'secondary',
      shape: 'storage'
    })
    expect(amazon).not.toHaveProperty('icon')
    expect(customer).toMatchObject({
      color: 'secondary',
      shape: 'storage'
    })
    expect(customer).not.toHaveProperty('icon')

    expect(cloud).toMatchObject({
      color: 'muted',
      shape: 'storage',
      icon: 'http://some-icon'
    })
    expect(frontend).toMatchObject({
      parent: 'cloud',
      color: 'secondary',
      shape: 'browser'
    })
    expect(frontend).not.toHaveProperty('icon')
  })

  it('should include by element kind', () => {
    const { nodeIds, edgeIds } = computeView({
      include: [
        {
          elementKind: 'system' as ElementKind,
          isEqual: true
        }
      ]
    })

    expect(nodeIds).toEqual(['cloud', 'email', 'amazon'])
    expect(edgeIds).toEqual([
      'cloud:amazon',
      'cloud:email',
      'email:cloud'
    ])
  })

  it('should include by element tag', () => {
    const { nodeIds, edgeIds } = computeView({
      include: [
        {
          elementTag: 'old' as Tag,
          isEqual: true
        }
      ]
    })

    expect(nodeIds).toEqual(['cloud', 'cloud.backend.storage', 'cloud.frontend.adminPanel'])

    expect(edgeIds).toEqual([])
  })

  it('should exclude by element tag and kind', () => {
    const { nodeIds, edgeIds } = computeView('cloud', [
      $include('*'),
      $include('cloud.backend.*'),
      $include('cloud.frontend.*'),
      $exclude('email'),
      {
        exclude: [
          {
            elementKind: 'actor' as ElementKind,
            isEqual: true
          },
          {
            elementTag: 'old' as Tag,
            isEqual: true
          }
        ]
      }
    ])

    expect(nodeIds).toEqual([
      'cloud.frontend',
      'cloud.frontend.dashboard',
      'cloud.backend',
      'amazon',
      'cloud.backend.graphql'
    ])

    expect(edgeIds).to.have.same.members([
      'cloud.frontend.dashboard:cloud.backend.graphql',
      'cloud.backend:amazon'
    ])
  })
})
