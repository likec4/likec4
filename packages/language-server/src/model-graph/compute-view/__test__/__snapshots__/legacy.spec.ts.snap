// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compute-element-view > view of cloud 1`] = `
{
  "__": "element",
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "id": "customer:cloud.frontend",
      "label": "opens in browser",
      "parent": null,
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend",
    },
    {
      "id": "support:cloud.frontend",
      "label": "manages",
      "parent": null,
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud.frontend",
    },
    {
      "id": "cloud.frontend:cloud.backend",
      "label": "requests",
      "parent": "cloud",
      "relations": [
        "cloud.frontend:cloud.backend",
        "cloud.frontend.dashboard:cloud.backend.graphql",
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend",
      "target": "cloud.backend",
    },
    {
      "id": "cloud.backend:amazon",
      "label": "uploads",
      "parent": null,
      "relations": [
        "cloud.backend.storage:amazon.s3",
      ],
      "source": "cloud.backend",
      "target": "amazon",
    },
    {
      "id": "cloud.backend:email",
      "label": "schedule",
      "parent": null,
      "relations": [
        "cloud.backend:email",
      ],
      "source": "cloud.backend",
      "target": "email",
    },
  ],
  "hash": "4b0570a97ddeaf4745e3affb567130c055a98a01",
  "id": "index",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "customer",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "outEdges": [
        "customer:cloud.frontend",
      ],
      "parent": null,
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "customer",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "support",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "outEdges": [
        "support:cloud.frontend",
      ],
      "parent": null,
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "support",
    },
    {
      "children": [
        "cloud.frontend",
        "cloud.backend",
      ],
      "color": "primary",
      "depth": 1,
      "description": null,
      "id": "cloud",
      "inEdges": [
        "support:cloud.frontend",
        "customer:cloud.frontend",
      ],
      "kind": "system",
      "level": 0,
      "links": null,
      "outEdges": [
        "cloud.backend:amazon",
        "cloud.backend:email",
      ],
      "parent": null,
      "shape": "rectangle",
      "style": {},
      "tags": [
        "next",
        "old",
      ],
      "technology": null,
      "title": "cloud",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend",
      "inEdges": [
        "support:cloud.frontend",
        "customer:cloud.frontend",
      ],
      "kind": "container",
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.frontend:cloud.backend",
      ],
      "parent": "cloud",
      "shape": "browser",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "frontend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend",
      "inEdges": [
        "cloud.frontend:cloud.backend",
      ],
      "kind": "container",
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.backend:amazon",
        "cloud.backend:email",
      ],
      "parent": "cloud",
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "backend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "email",
      "inEdges": [
        "cloud.backend:email",
      ],
      "kind": "system",
      "level": 0,
      "links": null,
      "outEdges": [],
      "parent": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "email",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "amazon",
      "inEdges": [
        "cloud.backend:amazon",
      ],
      "kind": "system",
      "level": 0,
      "links": null,
      "outEdges": [],
      "parent": null,
      "shape": "rectangle",
      "style": {},
      "tags": [
        "aws",
      ],
      "technology": null,
      "title": "amazon",
    },
  ],
  "tags": null,
  "title": null,
  "viewOf": "cloud",
}
`;

exports[`compute-element-view > view of cloud.frontend 1`] = `
{
  "__": "element",
  "autoLayout": "TB",
  "description": null,
  "edges": [
    {
      "id": "support:cloud.frontend.adminPanel",
      "label": "manages",
      "parent": null,
      "relations": [
        "support:cloud.frontend.adminPanel",
      ],
      "source": "support",
      "target": "cloud.frontend.adminPanel",
    },
    {
      "id": "customer:cloud.frontend.dashboard",
      "label": "opens in browser",
      "parent": null,
      "relations": [
        "customer:cloud.frontend.dashboard",
      ],
      "source": "customer",
      "target": "cloud.frontend.dashboard",
    },
    {
      "id": "cloud.frontend.adminPanel:cloud.backend",
      "label": "fetches",
      "line": "dashed",
      "parent": null,
      "relations": [
        "cloud.frontend.adminPanel:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.adminPanel",
      "tail": "odiamond",
      "target": "cloud.backend",
    },
    {
      "id": "cloud.frontend.dashboard:cloud.backend",
      "label": "requests",
      "line": "solid",
      "parent": null,
      "relations": [
        "cloud.frontend.dashboard:cloud.backend.graphql",
      ],
      "source": "cloud.frontend.dashboard",
      "target": "cloud.backend",
    },
  ],
  "hash": "5cd9190303add83e7131f7cf678d24e06f95f08b",
  "id": "index",
  "links": null,
  "nodes": [
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "support",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "outEdges": [
        "support:cloud.frontend.adminPanel",
      ],
      "parent": null,
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "support",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "customer",
      "inEdges": [],
      "kind": "actor",
      "level": 0,
      "links": null,
      "outEdges": [
        "customer:cloud.frontend.dashboard",
      ],
      "parent": null,
      "shape": "person",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "customer",
    },
    {
      "children": [
        "cloud.frontend.adminPanel",
        "cloud.frontend.dashboard",
      ],
      "color": "primary",
      "depth": 1,
      "description": null,
      "id": "cloud.frontend",
      "inEdges": [
        "customer:cloud.frontend.dashboard",
        "support:cloud.frontend.adminPanel",
      ],
      "kind": "container",
      "level": 0,
      "links": null,
      "outEdges": [
        "cloud.frontend.dashboard:cloud.backend",
        "cloud.frontend.adminPanel:cloud.backend",
      ],
      "parent": null,
      "shape": "browser",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "frontend",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend.adminPanel",
      "inEdges": [
        "support:cloud.frontend.adminPanel",
      ],
      "kind": "component",
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.frontend.adminPanel:cloud.backend",
      ],
      "parent": "cloud.frontend",
      "shape": "rectangle",
      "style": {},
      "tags": [
        "old",
      ],
      "technology": null,
      "title": "adminPanel",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.frontend.dashboard",
      "inEdges": [
        "customer:cloud.frontend.dashboard",
      ],
      "kind": "component",
      "level": 1,
      "links": null,
      "outEdges": [
        "cloud.frontend.dashboard:cloud.backend",
      ],
      "parent": "cloud.frontend",
      "shape": "rectangle",
      "style": {},
      "tags": [
        "next",
      ],
      "technology": null,
      "title": "dashboard",
    },
    {
      "children": [],
      "color": "primary",
      "description": null,
      "id": "cloud.backend",
      "inEdges": [
        "cloud.frontend.dashboard:cloud.backend",
        "cloud.frontend.adminPanel:cloud.backend",
      ],
      "kind": "container",
      "level": 0,
      "links": null,
      "outEdges": [],
      "parent": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "backend",
    },
  ],
  "tags": null,
  "title": null,
  "viewOf": "cloud.frontend",
}
`;
